### PR TITLE
1 0 recording refactor

### PR DESCRIPTION
**Pull request by ronyad**

### CODE REVIEW

File: recordingUploader/ConcatinationTask.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    @staticmethod
+    def sorted_ls(path):
+        mtime = lambda f: os.stat(os.path.join(path, f)).st_mtime
+        return list(sorted(os.listdir(path), key=mtime))
+
+    def create_manifest(self):
+        self.logger.info("Creating manifest file for the recorded entry %s", self.entry_id)
+        full_path = os.path.join(self.recording_path, self.manifest_input_file)
+        fo = open(full_path, "wb")
```
Comment (kalturaguy) [inline recordingUploader/ConcatinationTask.py]: use with open(full_path, "wb") as fo:

File: recordingUploader/TaskRunner.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+                self.logger.error("Failed to perform task :%s \n %s", str(e), traceback.format_exc())
+                retries = self.get_retry_count(src)
+                if retries > 0:
+                    self.logger.info("Job %s on entry %s has %s retries, move it to failed task directory ",
+                                     self.task_name, task_parameter['directory'], retries)
+                    shutil.move(src, self.failed_tasks_directory)
+                else:
+                    self.logger.fatal("Job %s on entry %s has no more retries or failed to get it, move entry to "
+                                      "failed task directory ", self.task_name, task_parameter['directory'])
+                    shutil.move(src, self.error_directory)
```
Comment (kalturaguy) [inline recordingUploader/TaskRunner.py]: wrap with try-except

File: recordingUploader/TaskRunner.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    def get_retry_count(self, src):
+        try:
+            retries_file_path = os.path.join(src, 'retries')
+            if not os.path.exists(retries_file_path):
+                retries_file = open(retries_file_path, "w")
```
Comment (kalturaguy) [inline recordingUploader/TaskRunner.py]: use with

File: recordingUploader/TaskRunner.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            self.logger.fatal("Failed to start task runner: %s  \n %s ", str(e), traceback.format_exc())
+
+
+class TaskBase(object):
+
+    hostname = gethostname()
+    base_directory = os.path.join(get_config('recording_base_dir'), hostname)
+
+    def check_stamp(self):
+        stamp_file = open(self.stamp_full_path, "r")  # w+ since we truncated the file
```
Comment (kalturaguy) [inline recordingUploader/TaskRunner.py]: with open

File: recordingUploader/TaskRunner.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        self.logger = logging.getLogger(logger_info)
+        self.output_filename = self.entry_directory+'_out.mp4'
+        self.recording_path = os.path.join(self.base_directory, self.__class__.__name__, 'processing',
+                                           self.entry_directory)
+        self.stamp_full_path = os.path.join(self.recording_path, 'stamp')
+
+    def write_stamp(self):
+
+        self.logger.info("About to write stamp %s on %s", self.timestamp, self.recording_path)
+        stamp_file = open(self.stamp_full_path, "w+")  # w+ since we truncated the file
```
Comment (kalturaguy) [inline recordingUploader/TaskRunner.py]: use with

File: recordingUploader/TaskRunner.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        stamp = stamp_file.read()
+        if stamp == self.timestamp:
+            self.logger.debug("Stamp  %s is not changed", stamp)
+        else:
+            msg = "Stamps are not equal! process stamp:%s, found in file: %s, abort directory" % (self.timestamp, stamp)
+            retries_file_path = os.path.join(self.recording_path, 'retries')
+            retries_file = open(retries_file_path, "w+")
+            retries_file.write('0')
+            retries_file.close()
+            raise ValueError(msg)
```
Comment (kalturaguy) [inline recordingUploader/TaskRunner.py]: create a typed exception

File: recordingUploader/UploadTask.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            upload_token_list_response = backend_client.upload_token_list(self.partner_id, file_name)
+            if upload_token_list_response.totalCount == 0:
+                self.token_id = backend_client.upload_token_add(self.partner_id, file_name, file_size)
+                self.uploaded_file_size = 0
+                return
+            if upload_token_list_response.totalCount == 1:  # if token is exist
+                self.token_id = upload_token_list_response.objects[0].id
+                self.uploaded_file_size = upload_token_list_response.objects[0].uploadedFileSize
+                UploadTask.logger.info("Found token exist for %s, token: %s, stating from %s", self.file_name, self.token_id,
+                                 self.uploaded_file_size)
```
Comment (kalturaguy) [inline recordingUploader/UploadTask.py]: indention problem

File: recordingUploader/recording_logger.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    log_full_path = get_config('log_file_name')
+    dictConfig(DEFAULT_LOGGING)
+
+    logging.basicConfig(
+        filename=log_full_path,
+        level=logging.DEBUG,
+        format='[%(asctime)s.%(msecs)d][%(process)d/%(threadName)s] [%(levelname)s] [%(name)s]: [%(funcName)s():%(lineno)s] %(message)s',
+        datefmt='%Y-%m-%d %H:%M:%S'
+    )
+    file_handler = logging.handlers.TimedRotatingFileHandler(log_full_path, when="m", interval=1)
```
Comment (kalturaguy) [inline recordingUploader/recording_logger.py]: change per day, also check for gzip

File: common/config/config.json.template [hunk 9]
```diff
@@ -2,6 +2,11 @@
   "mockNetwork" : false,
   "mockBackend" : false,
   "readMockBackendResponseFromFile" : false,
+  "recordingFolderPath" : "~/dvr/recording/Open_sessions",
+  "completedRecordingFolderPath" : "~/dvr/recording/ConcatenationTask/incoming",
+  "recordingTimeInterval" : "60000",
+  "recordingMaxDuration" : "86400",
+  "recordingSessionDuration" : "180000",
```
Comment (ronyad) [inline common/config/config.json.template]: add recording to seperate key, and change to second.

File: lib/playlistGenerator/PlaylistGenerator.js [hunk 10]
```diff
@@ -49,14 +49,35 @@ function PlaylistGenerator (entryObject) {
     that.stopPromise = null;
     that.startPromise = null;
     that.persistPromise = qResolved;
-    that.playListLimits =  {
-        manifestTimeWindowInMsec: Math.ceil(entryObject.playWindow * 1000 * hlsStoreFilefactor),
-        maxClipCountPerPlaylist: Math.max(10,entryObject.maxChunkCount)
-    };
-    that.playlistPath = path.join(persistenceFormat.getEntryBasePath(that.entryId),persistenceFormat.getMasterManifestName());
+    if (options && 'recording' in options && options.recording == true){
+        that.playListLimits =  {
```
Comment (ronyad) [inline lib/playlistGenerator/PlaylistGenerator.js]: use configuration

File: lib/playlistGenerator/ConcatSource.js [hunk 11]
```diff
@@ -169,7 +169,7 @@ ConcatSource.prototype.checkExpires = function (clock,totalDuration) {
             popClip.call(that);
         }
     }
-    else if (that.playlist.playListLimits.maxClipCountPerPlaylist) {
+    else if (that.playlist.playListLimits.maxClipCountPerPlaylist && that.playlist.playListLimits.maxClipCountPerPlaylist != 'INF') {
```
Comment (ronyad) [inline lib/playlistGenerator/ConcatSource.js]: check to write null instead of INF

File: lib/playlistGenerator/Playlist.js [hunk 12]
```diff
@@ -35,7 +35,7 @@ var machineName = hostname.getLocalMachineFullHostname();
 *   sequences : (collection of Clip) flavor collection containing MixFilterClip objects
 *   durations:  (collection of int) total clip durations; actually it's always 1 long since we use MixFilterClip. along with firstClipTime describes playlist window.
 * */
-function Playlist(loggerInfo, serializationCtx) {
+function Playlist(loggerInfo, serializationCtx, playlistType) {
```
Comment (ronyad) [inline lib/playlistGenerator/Playlist.js]: check to remove playlistType

File: lib/Controller.js [hunk 13]
```diff
@@ -239,10 +240,9 @@ module.exports = (function(){
         var that = this;
         logger.warn("Starting liveProxy pid: %s, wowza: '%s', API server: '%s', prefix='%s'",process.pid,config.get("mediaServer").hostname,config.get('backendClient').serviceUrl,that.prefix);
         logger.warn("Log files %s",config.get('logFileName'));
-        return Q.fcall(function() {
-            that._intervalObj = setInterval(handleEntries.bind(that), that.pollingInterval);
-            return handleEntries.call(that);
-        });
+        that._intervalObj = setInterval(handleEntries.bind(that), that.pollingInterval);
```
Comment (kalturaguy) [inline lib/Controller.js]: ??

File: lib/RecordingManager.js [hunk 14]
```diff
@@ -0,0 +1,531 @@
+            /**
```
Comment (kalturaguy) [inline lib/RecordingManager.js]: identation problem

File: lib/RecordingManager.js [hunk 15]
```diff
@@ -0,0 +1,531 @@
...
+                        return Q.resolve(resultsValueSorted) // return resolve anyway!
+                    })
+            }
+
+                LinkAndUpdateJson(Mp4Files) {
+                    var promises = _.map(Mp4Files, function (Mp4File) {
+                        let SourcefilePath = Mp4File.path;
+                        let chunkName = path.basename(Mp4File.path)
+                        let arrayPath = Mp4File.path.split(path.sep);
+                        let flavorId = arrayPath[arrayPath.length - 3];
```
Comment (kalturaguy) [inline lib/RecordingManager.js]: pass as argument

File: lib/RecordingManager.js [hunk 16]
```diff
@@ -0,0 +1,531 @@
...
+                        })
+                    })
+                }
+
+                /*
+                 In order to check that all chunks in live directory exist in recording, we check first that all chunks in recording are in json
+                 */
+                restoreSession() {
+
+                    return this.accessibleRecording() //in order to check if file exist, first need to load Json form disk
```
Comment (kalturaguy) [inline lib/RecordingManager.js]: split accesiableRecording to common code

File: lib/entry/LiveEntry.js [hunk 17]
```diff
@@ -112,8 +114,13 @@ LiveEntry.prototype.start = function() {
             // Create master manifest and save it on disk
             return that.playlistGenerator.finalizeStart();
         })
-        .then(function(flavorsObjArray) {
+        .then(function (flavorsObjArray) {
             that.flavorsObjArray = flavorsObjArray;
+            if (that.entryObject.recordStatus !== kalturaTypes.KalturaRecordStatus.DISABLED  && that.entryObject.serverType === kalturaTypes.KalturaEntryServerNodeType.LIVE_PRIMARY ) { //if recording is enable
+                return  RecordingManager.startRedording(that.entryObject, that.flavorsObjArray) //todo promise? its not depending on nothing !!! so dont wait faild ? what to do ?
```
Comment (kalturaguy) [inline lib/entry/LiveEntry.js]: misspelling Redording

File: common/config/config.json.template [hunk 18]
```diff
@@ -23,7 +23,7 @@
   "logToConsole" : false,
   "logFileName" : "~/dvr/kLiveController.log",
   "logTimeZoneOffset" : 300,
-  "monitorPort": 4390,
+  "monitorPort": 4391,
```
Comment (kalturaguy) [inline common/config/config.json.template]: Y?

File: common/config/config.json.template [hunk 19]
```diff
@@ -64,5 +64,12 @@
     "preserveMP4Chunks": false,
     "segmentDuration" : 10000,
     "enablePlaylistHistory": false
-  }
+  },
+  "recording" : {
+       "recordingFolderPath" : "/web/content/kLive/recording/recordings",
```
Comment (kalturaguy) [inline common/config/config.json.template]: it's better if you put ther ~/dvr/recording

File: common/config/config.json.template [hunk 20]
```diff
@@ -64,5 +64,12 @@
...
+       "completedRecordingFolderPath" : "/web/content/kLive/recording/incoming",
+       "recordingTimeInterval" : "60000",
```
Comment (kalturaguy) [inline common/config/config.json.template]: please change it to recordingTimeIntervalInSeconds: 60

File: common/config/config.json.template [hunk 21]
```diff
@@ -64,5 +64,12 @@
...
+       "recordingMaxDuration" : "86400",
+       "recordingSessionDuration" : "180000"
```
Comment (kalturaguy) [inline common/config/config.json.template]: change recordingSessionDurationInSeconds: 180

File: lib/Controller.js [hunk 22]
```diff
@@ -240,8 +241,8 @@ module.exports = (function(){
         logger.warn("Starting liveProxy pid: %s, wowza: '%s', API server: '%s', prefix='%s'",process.pid,config.get("mediaServer").hostname,config.get('backendClient').serviceUrl,that.prefix);
         logger.warn("Log files %s",config.get('logFileName'));
         return Q.fcall(function() {
-            that._intervalObj = setInterval(handleEntries.bind(that), that.pollingInterval);
-            return handleEntries.call(that);
+        that._intervalObj = setInterval(handleEntries.bind(that), that.pollingInterval);
```
Comment (kalturaguy) [inline lib/Controller.js]: fix indentation

File: lib/recording/RecordingEntrySession.js [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    LinkAndUpdateJson(Mp4Files) {
+        let Mp4FilesPath = _.map(Mp4Files, function(element){ return  element.path; });
+        this.logger.debug("LinkAndUpdateJson : %s", Mp4FilesPath)
+        var promises = _.map(Mp4Files, (Mp4File) =>{
+            let SourcefilePath = Mp4File.path;
+            let chunkName = path.basename(Mp4File.path)
+            let arrayPath = Mp4File.path.split(path.sep);
+            let flavorId = arrayPath[arrayPath.length - 3];
```
Comment (kalturaguy) [inline lib/recording/RecordingEntrySession.js]: check if you can get rid of flavorId

File: common/config/config.json.template (hunk 19 above)
Comment (kalturaguy) [inline common/config/config.json.template]: add enable: "False" by default

File: common/config/config.json.template (hunk 18 above)
Comment (ronyad) [inline common/config/config.json.template]: done

File: common/config/config.json.template (hunk 19 above)
Comment (ronyad) [inline common/config/config.json.template]: done

File: common/config/config.json.template (hunk 20 above)
Comment (ronyad) [inline common/config/config.json.template]: done

File: lib/Controller.js (hunk 22 above)
Comment (ronyad) [inline lib/Controller.js]: done

File: lib/MP4WriteStream.js [hunk 24]
```diff
@@ -167,6 +167,10 @@ class MP4File {
     get path(){
         return this[_stream].chunkPath;
     }
+
```
Comment (kalturaguy) [inline lib/MP4WriteStream.js]: duplicated

File: lib/entry/FlavorDownloader.js [hunk 25]
```diff
@@ -43,6 +44,8 @@ var FlavorDownloader = (function() {
         this.id = this.constructor.uniqueId;
         this.constructor.uniqueId = this.constructor.uniqueId + 1;
         this.entryId = entryObj.entryId;
+        this.recordStatus = entryObj.recordStatus  &&
```
Comment (kalturaguy) [inline lib/entry/FlavorDownloader.js]: this.recordingEnabled

File: packager/config/nginx.conf.template [hunk 26]
```diff
@@ -146,8 +152,11 @@ http {
 
         # fetch files from CONTENT_DIR
         location / {
-            alias @CONTENT_DIR@;
+            alias @CONTENT_DIR@/$entryId/playlist.json;
```
Comment (kalturaguy) [inline packager/config/nginx.conf.template]: please revert

File: recordingUploader/BackendClient.py [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+    type = KalturaSessionType.ADMIN
+
+    def __init__(self, session_id):
+        self.logger = logger_decorator(self.__class__.__name__, session_id)
+
+    def create_new_session(self):
+        result = self.client.session.start(self.admin_secret, None, self.type, self.partner_id, None, None)
+        BackendClient.ks = result[0]
+        BackendClient.expiration_time_ks = int(self.session_duration) + int(time.time()) - 3600  # confidence interval
+        self.client.setKs(self.ks)
```
Comment (kalturaguy) [inline recordingUploader/BackendClient.py]: did you mean BackendClient.ks ?

File: recordingUploader/BackendClient.py [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+        self.logger.info("Creating a new session, KS= %s \n Header: %s", self.ks, result[1])
+
+    def get_kaltura_session(self):
+        self.mutex.acquire()
+        try:
+            if (self.ks is None) or self.expiration_time_ks < int(time.time()):
```
Comment (kalturaguy) [inline recordingUploader/BackendClient.py]: BackendClient.ks

File: recordingUploader/Config/config.ini [hunk 29]
```diff
@@ -0,0 +1,26 @@
+[default_config]
+max_task_count=10
+recording_base_dir = /web/content/kLive/recording
+ffmpeg_path = /opt/kaltura/ffmpeg
```
Comment (kalturaguy) [inline recordingUploader/Config/config.ini]: remove ffmpeg

File: recordingUploader/BackendClient.py [hunk 30]
```diff
@@ -0,0 +1,159 @@
...
+    client = KalturaClient(config)
+    client.setPartnerId(partner_id)
+    ks = None
+    type = KalturaSessionType.ADMIN
+
+    def __init__(self, session_id):
+        self.logger = logger_decorator(self.__class__.__name__, session_id)
+
+    def create_new_session(self):
+        result = self.client.session.start(self.admin_secret, None, self.type, self.partner_id, None, None)
```
Comment (kalturaguy) [inline recordingUploader/BackendClient.py]: generate ks without calling api

File: recordingUploader/Tasks/ConcatinationTask.py [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+        regex = r"(.*.ts)$"
+        matches = re.findall(regex, m3u8, re.MULTILINE)
+        return matches
+
+    def run(self):
+
+        self.write_stamp()
+        output_full_path = os.path.join(self.recording_path, self.output_filename)
+        m3u8 = self.download_file(self.url_source_manifest)
+        self.logger.debug("load recording manifest : %s ", m3u8)
```
Comment (kalturaguy) [inline recordingUploader/Tasks/ConcatinationTask.py]: load**ed**

File: recordingUploader/Tasks/UploadTask.py [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+                continue
+            final_chunk = sequence_number == chunks_to_upload
+            resume = sequence_number > 1
+
+            chunk = UploadChunkJob(upload_session, infile, sequence_number, final_chunk, resume_at, resume, self.backend_client)
+
+            ThreadWorkers().add_job(chunk)
+
+        result = ThreadWorkers().wait_for_all_jobs_done()
+        self.check_stamp()
```
Comment (kalturaguy) [inline recordingUploader/Tasks/UploadTask.py]: 1. should close infile
2. should protect closing of infile with try,except

File: lib/recording/RecordingManager.js [hunk 33]
```diff
@@ -0,0 +1,180 @@
...
+            })
+            .catch((err) => {
+                logger.error(ErrorUtils.error2string(err))
+            })
+
+}
+
+
+}
+if (configRecording.enable){
```
Comment (kalturaguy) [inline lib/recording/RecordingManager.js]: why not do 
if (!configRecording.enable) {
return;
}

File: lib/playlistGenerator/PlaylistGenerator.js [hunk 34]
```diff
@@ -52,10 +52,34 @@ function PlaylistGenerator (entryObject) {
...
+                    return Q.resolve(options.flavors)
+                }
+            }
+        }
+
+    }
+    else{
+        that.playListLimits =  {
+            manifestTimeWindowInMsec: Math.ceil(entryObject.playWindow * 1000 * hlsStoreFilefactor),
+            maxClipCountPerPlaylist: Math.max(10,entryObject.maxChunkCount)
```
Comment (igorshevach) [inline lib/playlistGenerator/PlaylistGenerator.js]: maxClipCountPerPlaylist does not exist any more please delete this line

File: lib/playlistGenerator/PlaylistGenerator.js [hunk 35]
```diff
@@ -94,6 +118,7 @@ var setPlayList = function (playlist){
         that.playlistImp.inner.liveWindowDuration = Math.ceil(that.playListLimits.manifestTimeWindowInMsec / hlsStoreFilefactor);
     }
     that.playlistImp.addListener(that);
+    that.playlistImp.inner.presentationEndTime = 0
```
Comment (igorshevach) [inline lib/playlistGenerator/PlaylistGenerator.js]: if(live) 
that.playlistImp.inner.presentationEndTime = that.playlistImp.getCurrentTime() + MILLISEC_IN_YEAR * 10;
else 
that.playlistImp.inner.presentationEndTime = 0

File: lib/MP4WriteStream.js [hunk 36]
```diff
@@ -166,6 +168,10 @@ class MP4File {
...
+    set path(path){
+        this[_stream].chunkPath = path;
```
Comment (igorshevach) [inline lib/MP4WriteStream.js]: try object clone instead of modifying path property

**Pull request closed without merging**